### PR TITLE
Extended marks 2: Electric Boogalo

### PR DESCRIPTION
**Pull request by bfredl**
#5031 PR thread is too long, so lets open a new one. Tests passes locally after some small adjustments, lets hope for the best.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -148,3 +148,14 @@ See `LICENSE` for details.
 [Gentoo]: https://packages.gentoo.org/packages/app-editors/neovim
 
 <!-- vim: set tw=80: -->
+
```
Comment (bfredl) [inline README.md]: remove I guess.

File: runtime/doc/api.txt [hunk 2]
```diff
@@ -438,6 +438,36 @@ Example: create a float with scratch buffer: >
...
+Example:
+
+We will set an extmark at row 1, column three.
+
+`let g:mark_ns = nvim_create_namespace('myplugin')`
+`let g:mark_id = nvim_buf_set_mark(0, :g:mark_ns, 0, 1, 3)`
+
+Note: the mark id was randomly generated because we used an inital value of 0
+
+`echo nvim_buf_lookup_mark(0, g:mark_ns, g:mark_id)`
```
Comment (bfredl) [inline runtime/doc/api.txt]: change

File: test/functional/legacy/046_multi_line_regexps_spec.lua [hunk 3]
```diff
@@ -8,22 +8,22 @@ local expect = helpers.expect
 describe('multi-line regexp', function()
   setup(clear)
 
-  it('is working', function()
+  it('is working #fail', function()
     insert([[
-      1 aa
```
Comment (bfredl) [inline test/functional/legacy/046_multi_line_regexps_spec.lua]: this should have been done by insert itself?

File: test/functional/ui/inccommand_spec.lua [hunk 4]
```diff
@@ -131,7 +131,7 @@ describe(":substitute, inccommand=split interactivity", function()
   end)
 end)
 
-describe(":substitute, 'inccommand' preserves", function()
+describe(":substitute, 'inccommand' preserves #inc", function()
```
Comment (bfredl) [inline test/functional/ui/inccommand_spec.lua]: old debug stuff? can already use `TEST_FILE=test/functional/ui/inccommand_spec.lua`

File: test/functional/ui/mouse_spec.lua [hunk 5]
```diff
@@ -817,7 +817,6 @@ describe('ui/mouse/input', function()
       feed_command('syntax match NonText "cats" conceal cchar=X')
       feed_command('syntax match NonText "x" conceal cchar=>')
 
-      -- First column is there to retain the tabs.
```
Comment (bfredl) [inline test/functional/ui/mouse_spec.lua]: accident?

File: src/nvim/ops.c [hunk 6]
```diff
@@ -126,6 +127,30 @@ static char opchars[][3] =
   { Ctrl_X, NUL, false },    // OP_NR_SUB
 };
 
+char *nvim_lltoa(int64_t val, int base)
```
Comment (bfredl) [inline src/nvim/ops.c]: Delete. only used for counting length, we can have a static function just for that for now.

File: src/nvim/api/private/helpers.c [hunk 7]
```diff
@@ -1505,3 +1507,129 @@ ArrayOf(Dictionary) keymap_array(String mode, buf_T *buf)
 
   return mappings;
 }
+
+// Returns an extmark given an id or a positional index
+// If throw == true then an error will be raised if nothing
+// was found
+// Returns NULL if something went wrong
+ExtendedMark *extmark_from_id_or_pos(Buffer buffer,
```
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: not used, but keep anyway for future API usage?

File: src/nvim/api/private/helpers.c [hunk 8]
```diff
@@ -1505,3 +1507,129 @@ ArrayOf(Dictionary) keymap_array(String mode, buf_T *buf)
...
+bool ns_initialized(uint64_t ns)
+{
+  if (ns < 1) {
+    return false;
+  }
+  return ns < (uint64_t)next_namespace_id;
+}
+
+// Extmarks may be queried from position or name or even special names
+// in the future such as "cursor". This macro sets the line and col
```
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: function

File: src/nvim/api/buffer.c [hunk 9]
```diff
@@ -999,6 +1003,213 @@ ArrayOf(Integer, 2) nvim_buf_get_mark(Buffer buffer, String name, Error *err)
...
+///
+/// @param buffer The buffer handle
+/// @param ns_id a identifier returned previously with nvim_create_namespace
+/// @param id The extmark's id or 0 for next free id
+/// @param row The row to set the extmark to.
+/// @param col The column to set the extmark to.
+/// @param[out] err Details of an error that may have occurred
+/// @return the nsmark_id for a new mark, or 0 for an update
+Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id, Integer id,
+                             Integer line, Integer col, Error *err)
```
Comment (bfredl) [inline src/nvim/api/buffer.c]: should add  `Dictionary options` and force it to be empty, for extensibility.

File: src/nvim/ex_cmds.c [hunk 10]
```diff
@@ -105,6 +106,8 @@ typedef struct {
 
 #ifdef INCLUDE_GENERATED_DECLARATIONS
 # include "ex_cmds.c.generated.h"
+#include "mark_extended.h"
```
Comment (bfredl) [inline src/nvim/ex_cmds.c]: noise?

File: src/nvim/map.h [hunk 11]
```diff
@@ -42,6 +42,7 @@ MAP_DECLS(handle_T, ptr_t)
 MAP_DECLS(String, MsgpackRpcRequestHandler)
 MAP_DECLS(HlEntry, int)
 MAP_DECLS(String, handle_T)
+MAP_DECLS(uint64_t, cstr_t)
```
Comment (bfredl) [inline src/nvim/map.h]: redundant, on RHS a string is just a pointer (probably not used here anyway anymore)

File: src/nvim/api/buffer.c [hunk 12]
```diff
@@ -999,6 +1003,220 @@ ArrayOf(Integer, 2) nvim_buf_get_mark(Buffer buffer, String name, Error *err)
...
+    ADD(rv, ARRAY_OBJ(mark));
+  }
+
+  kv_destroy(marks);
+  return rv;
+}
+
+/// Create or update a namespaced mark at a position
+///
+/// If an invalid namespace is given, an error will be raised.
```
Comment (bfredl) [inline src/nvim/api/buffer.c]: Write clearer that plugin passing `id=0` is expected when creating new marks (passing id for new mark should only be needed when "streaming" marks over RPC).

File: src/nvim/ops.c [hunk 13]
```diff
@@ -2694,6 +2763,34 @@ static void do_autocmd_textyankpost(oparg_T *oap, yankreg_T *reg)
 }
 
 
+
+// Function length couldn't be over 500 lines..
```
Comment (bfredl) [inline src/nvim/ops.c]: makes sense but remove comment

File: src/nvim/mark_extended.c [hunk 14]
```diff
@@ -0,0 +1,1262 @@
...
+// TODO(timeyyy): change mincol to be for the mark to be copied, not moved
+//
+// @param mincol First column that needs to be moved (start of delete range) + 1
+// @param endcol Last column which needs to be copied (end of delete range + 1)
+void extmark_col_adjust_delete(buf_T *buf, linenr_T lnum,
+                               colnr_T mincol, colnr_T endcol,
+                               ExtmarkOp undo, int _eol)
+{
+  colnr_T start_effected_range = mincol;
+  // TODO(timeyyy): understand why this has to be uncommented out for the
```
Comment (bfredl) [inline src/nvim/mark_extended.c]: appears fixed, delete comment.

File: src/nvim/undo.c [hunk 15]
```diff
@@ -2828,6 +2849,9 @@ u_freeentries(
     u_freeentry(uep, uep->ue_size);
   }
 
+  // TODO(timeyyy): is this the correct place? ...
```
Comment (bfredl) [inline src/nvim/undo.c]: yes (delete comment)

File: src/nvim/mark_extended.c [hunk 16]
```diff
@@ -0,0 +1,1262 @@
...
+                 0L,
+                 kExtmarkNoUndo,
+                 true);
+  }
+}
+
+// for anything other than deletes
+// Return, desired col amount where the adjustment should take place
+// (not taking) eol into account
+static long update_constantly(colnr_T _, colnr_T __, long col_amount)
```
Comment (bfredl) [inline src/nvim/mark_extended.c]: overengineering, just use a bool flag to `extmark_col_adjust_impl`

File: src/nvim/mark_extended.c [hunk 17]
```diff
@@ -0,0 +1,1262 @@
...
+  linenr_T line1 = undo_info.data.move.line1;
+  linenr_T line2 = undo_info.data.move.line2;
+  linenr_T last_line = undo_info.data.move.last_line;
+  linenr_T dest = undo_info.data.move.dest;
+  linenr_T num_lines = undo_info.data.move.num_lines;
+  linenr_T extra = undo_info.data.move.extra;
+
+  if (undo) {
+    if (dest >= line2) {
+      extmark_adjust(curbuf,
```
Comment (bfredl) [inline src/nvim/mark_extended.c]: style/nit: less linebreaks

File: test/functional/api/mark_extended_spec.lua [hunk 18]
```diff
@@ -0,0 +1,1367 @@
...
+    -- testing marks below are also moved
+    feed("yyP")
+    set_extmark(ns, marks[1], 0, 4)
+    set_extmark(ns, marks[2], 1, 4)
+    feed('1Gla<cr><esc>')
+    check_undo_redo(ns, marks[1], 0, 4, 1, 2)
+    check_undo_redo(ns, marks[2], 1, 4, 2, 4)
+  end)
+
+  -- TODO mark_col_adjust for normal marks fails in vim/neovim
```
Comment (bfredl) [inline test/functional/api/mark_extended_spec.lua]: not a regression so not an issue, improving "normal marks" is out of scope.

File: src/nvim/mark_extended.c [hunk 19]
```diff
@@ -0,0 +1,1262 @@
...
+
+  bool marks_exist = false;
+  linenr_T *lp;
+
+  linenr_T adj_start = line1;
+  if (amount == MAXLNUM) {
+    // Careful! marks from deleted region can end up on en extisting extline
+    // that is goinig to be adjusted to the target position.
+    linenr_T join_num = line1 - amount_after;
+    ExtMarkLine *joinline = join_num > line2 \
```
Comment (bfredl) [inline src/nvim/mark_extended.c]: style

**Comment by bfredl** [comment]
I've done some revisions, might do a few more (like a few more obsolete TODO comments and such) but this is pretty much mergable from my point of view. We have the 0.5 cycle to fix bugs/edge cases (which will be reported by master testers after merge :) . This is is pretty much "do nothing unless invoked", so it should not affect stability in general.

Docs can and will be improved after merge (but of course point out already if there is something obviously wrong/misleading).

I also plan to heavily refactor the internal implementation based on #9404 (a more efficient kbtree that don't need two separate layers for columns/lines) at some point, but this is not critical. There is no catastrophe if this happens after 0.5 (or never).

File: runtime/doc/api.txt [hunk 20]
```diff
@@ -438,6 +438,36 @@ Example: create a float with scratch buffer: >
...
+Buffer extended marks	                                  *api-extended-marks*
+
+An extended mark represents a buffer annotation that remains logically
+stationary even as the buffer changes. They could be used to represent cursors,
+folds, misspelled words, and anything else that needs to track a logical
+location in the buffer over time.
+
+Example:
+
+We will set an extmark at row 1, column three.
```
Comment (zgpio) [inline runtime/doc/api.txt]: `1` and `three` are inconsistent

File: runtime/doc/api.txt [hunk 21]
```diff
@@ -438,6 +438,36 @@ Example: create a float with scratch buffer: >
...
+
+`let g:mark_ns = nvim_create_namespace('myplugin')`
+`let g:mark_id = nvim_buf_set_extmark(0, :g:mark_ns, 0, 1, 3)`
```
Comment (zgpio) [inline runtime/doc/api.txt]: Is `:g:mark_ns` a typo?

File: runtime/doc/api.txt [hunk 22]
```diff
@@ -438,6 +438,36 @@ Example: create a float with scratch buffer: >
...
+
+Note: the mark id was generated because we used an inital value of 0
+
+`echo nvim_buf_get_extmark_by_id(0, g:mark_ns, g:mark_id)`
+=> [1, 1, 3]
```
Comment (zgpio) [inline runtime/doc/api.txt]: Is this example outdated？

**Comment by bfredl** [comment]
@zgpio updated to your suggestions (and a bit more).

File: runtime/doc/api.txt [hunk 23]
```diff
@@ -438,6 +438,42 @@ Example: create a float with scratch buffer: >
...
+indexed, use row and column counts 0 and 2:
+
+`let g:mark_ns = nvim_create_namespace('myplugin')`
+`let g:mark_id = nvim_buf_set_extmark(0, g:mark_ns, 0, 0, 2)`
+
+Passing in id=0 creates a new mark and returns the id. we can look-up a mark
+by its id:
+
+`echo nvim_buf_get_extmark_by_id(0, g:mark_ns, g:mark_id)`
+=> [1, 0, 2]
```
Comment (zgpio) [inline runtime/doc/api.txt]: According to `nvim_buf_get_extmark_by_id`, the return value  should be `[0, 2]`.

Comment (bfredl) [reply runtime/doc/api.txt]: Thanks, updated.

File: runtime/doc/api.txt [hunk 24]
```diff
@@ -1747,6 +1783,78 @@ nvim_buf_get_mark({buffer}, {name})                      *nvim_buf_get_mark()*
...
+
+                all_marks = nvim_buf_get_extmarks(0, my_ns, 0, -1, -1)
+
+                If end is a lower position than start, then the range will be
+                traversed backwards. This is mostly used with limited amount,
+                to be able to get the first marks prior to a given position.
+
+                Parameters: ~
+                    {buffer}     The buffer handle
+                    {namespace}  An id returned previously from
```
Comment (bfredl) [inline runtime/doc/api.txt]: wrong parameter names (also other functions)

**Comment by justinmk** [comment]
Tried following the instructions (after making  some corrections). Created an extmark, then closed a window and got a segfault:

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00000000005a8f0a in extmark_free_all (buf=0x2331a30) at ../src/nvim/mark_extended.c:339
339       map_foreach(buf->b_extmark_ns, ns, ns_obj, {
[Current thread is 1 (Thread 0x7ff8304a2740 (LWP 7108))]
(gdb) bt
+bt
#0  0x00000000005a8f0a in extmark_free_all (buf=0x2331a30) at ../src/nvim/mark_extended.c:339
#1  0x000000000044e1b3 in free_buffer_stuff (buf=0x2331a30, free_flags=3) at ../src/nvim/buffer.c:820
#2  0x000000000044ceab in close_buffer (win=0x295e1b0, buf=0x2331a30, action=2, abort_if_last=1) at ../src/nvim/buffer.c:630
#3  0x00000000007466d5 in win_close (win=0x295e1b0, free_buf=true) at ../src/nvim/window.c:2534
#4  0x0000000000524b89 in ex_quit (eap=0x7ffcd49f5d68) at ../src/nvim/ex_docmd.c:6158
#5  0x00000000005144b5 in do_one_cmd (cmdlinep=0x7ffcd49f6108, flags=10, cstack=0x7ffcd49f6110, fgetline=0x0, cookie=0x0) at ../src/nvim/ex_docmd.c:2289
#6  0x0000000000510147 in do_cmdline (cmdline=0x7ffcd49f6960 "quit", fgetline=0x0, cookie=0x0, flags=10) at ../src/nvim/ex_docmd.c:596
#7  0x0000000000511094 in do_cmdline_cmd (cmd=0x7ffcd49f6960 "quit") at ../src/nvim/ex_docmd.c:270
#8  0x000000000074482e in do_window (nchar=17, Prenum=0, xchar=0) at ../src/nvim/window.c:173
#9  0x00000000005ea778 in nv_window (cap=0x7ffcd49f6ab8) at ../src/nvim/normal.c:6523
#10 0x00000000005e9764 in normal_execute (state=0x7ffcd49f6a30, key=23) at ../src/nvim/normal.c:1131
#11 0x00000000006d826d in state_enter (s=0x7ffcd49f6a30) at ../src/nvim/state.c:69
#12 0x00000000005e0ca4 in normal_enter (cmdwin=false, noexmode=false) at ../src/nvim/normal.c:463
#13 0x00000000005932b2 in main (argc=1, argv=0x7ffcd49f6e58) at ../src/nvim/main.c:577
```

File: src/nvim/mark_extended.c [hunk 25]
```diff
@@ -0,0 +1,1134 @@
...
+                    linenr_T line2,
+                    long amount,
+                    long amount_after,
+                    ExtmarkOp undo,
+                    bool end_temp)
+{
+  ExtMarkLine *_extline;
+
+  // btree needs to be kept ordered to work, so far only :move requires this
+  // 2nd call with end_temp =  unpack the lines from the temp position
```
Comment (justinmk) [inline src/nvim/mark_extended.c]: `end_temp = ?`

File: src/nvim/mark_extended.c [hunk 26]
```diff
@@ -0,0 +1,1134 @@
...
+void u_extmark_move(buf_T *buf, linenr_T line1, linenr_T line2,
+                    linenr_T last_line, linenr_T dest, linenr_T num_lines,
+                    linenr_T extra)
+{
+  u_header_T  *uhp = force_get_undo_header(buf);
+  if (!uhp) {
+    return;
+  }
+
+  AdjustMove move;
```
Comment (justinmk) [inline src/nvim/mark_extended.c]: what's the difference between `AdjustMove` and `Adjust` ? Should be a comment near the structs. Not to mention the other `FooAdjust` types...

File: src/nvim/mark_extended.h [hunk 27]
```diff
@@ -0,0 +1,267 @@
...
+    Adjust adjust;
+    ColAdjust col_adjust;
+    ColAdjustDelete col_adjust_delete;
+    AdjustMove move;
+    ExtmarkSet set;
+    ExtmarkUpdate update;
+    ExtmarkCopy copy;
+    ExtmarkCopyPlace copy_place;
+    ExtmarkClear clear;
+  } data;
```
Comment (justinmk) [inline src/nvim/mark_extended.h]: this seems extreme, I wonder if we could have just 1 `Adjust` type with various fields.

File: src/nvim/ops.c [hunk 28]
```diff
@@ -306,6 +307,20 @@ void shift_line(
     change_indent(INDENT_SET, count, false, NUL, call_changed_bytes);
   } else {
     (void)set_indent(count, call_changed_bytes ? SIN_CHANGED : 0);
+
+    colnr_T col_amount;
+    colnr_T mincol = (curwin->w_cursor.col + 1) -p_sw;
+    if (left) {
+      col_amount = -p_sw;
+    } else {
+      col_amount = p_sw;
```
Comment (justinmk) [inline src/nvim/ops.c]: ternary would be clearer and saves lines.

File: src/nvim/ops.c [hunk 29]
```diff
@@ -479,6 +494,13 @@ static void shift_block(oparg_T *oap, int amount)
   State = oldstate;
   curwin->w_cursor.col = oldcol;
   p_ri = old_p_ri;
+
+  colnr_T col_amount = p_sw;
+  if (left) {
+    col_amount = -col_amount;
+  }
```
Comment (justinmk) [inline src/nvim/ops.c]: ternary would be clearer and saves lines.

File: src/nvim/ops.c [hunk 30]
```diff
@@ -1632,10 +1665,41 @@ int op_delete(oparg_T *oap)
...
+    // + 1 to change to buf mode, then plus 1 to fit function requirements
+    endcol = oap->end.col + 1 + 1;
+
+    lnum = curwin->w_cursor.lnum;
+    if (oap->is_VIsual == false) {
+      // for some reason we required this :/
+      endcol = endcol - 1;
+      // for some reason we required this :/
+      if (endcol < mincol) {
+        endcol = mincol;
```
Comment (justinmk) [inline src/nvim/ops.c]: ternary

File: src/nvim/ops.c [hunk 31]
```diff
@@ -2694,6 +2763,32 @@ static void do_autocmd_textyankpost(oparg_T *oap, yankreg_T *reg)
...
+                            MotionType y_type,
+                            linenr_T lnum,
+                            colnr_T col)
+{
+  // adjust extmarks
+  colnr_T col_amount;
+  if (dir == FORWARD) {
+    col_amount = (colnr_T)(totlen-1);
+  } else {
+    col_amount = (colnr_T)totlen;
```
Comment (justinmk) [inline src/nvim/ops.c]: ternary

File: src/nvim/ops.c [hunk 32]
```diff
@@ -3756,19 +3859,26 @@ int do_join(size_t count,
     // If deleting more spaces than adding, the cursor moves no more than
     // what is added if it is inside these spaces.
     const int spaces_removed = (int)((curr - curr_start) - spaces[t]);
+    lnum = curwin->w_cursor.lnum + t;
+    mincol = (colnr_T)0;
+    lnum_amount = (linenr_T)-t;
+    col_amount = (long)(cend - newp - spaces_removed);
```
Comment (justinmk) [inline src/nvim/ops.c]: could declare here.

File: src/nvim/undo.c [hunk 33]
```diff
@@ -3022,3 +3045,28 @@ list_T *u_eval_tree(const u_header_T *const first_uhp)
 
   return list;
 }
+
+// Given the buffer, Return the undo header. If none is set, set one first.
+// NULL will be returned if e.g undolevels = -1 (undo disabled)
+u_header_T *force_get_undo_header(buf_T *buf)
```
Comment (justinmk) [inline src/nvim/undo.c]:
```suggestion
u_header_T *u_force_get_undo_header(buf_T *buf)
```

File: src/nvim/undo.c [hunk 34]
```diff
@@ -3022,3 +3045,28 @@ list_T *u_eval_tree(const u_header_T *const first_uhp)
...
+  }
+  // Create the first undo header for the buffer
+  if (!uhp) {
+    // TODO(timeyyy): there would be a better way to do this!
+    u_save_cursor();
+    uhp = buf->b_u_curhead;
+    if (!uhp) {
+      uhp = buf->b_u_newhead;
+      if (get_undolevel() > 0) {
+        assert(uhp);
```
Comment (justinmk) [inline src/nvim/undo.c]: `abort()`

File: src/nvim/mark_extended_defs.h [hunk 35]
```diff
@@ -0,0 +1,54 @@
...
+#define markitems_cmp(a, b) (extmark_cmp((a), (b)))
+KBTREE_INIT(markitems, ExtendedMark, markitems_cmp, 10)
+
+typedef struct ExtMarkLine
+{
+  linenr_T lnum;
+  kbtree_t(markitems) items;
+} ExtMarkLine;
+
+#define extline_cmp(a, b) (kb_generic_cmp((a)->lnum, (b)->lnum))
```
Comment (justinmk) [inline src/nvim/mark_extended_defs.h]: these macros should be uppercase, or inline functions.

File: src/nvim/mark_extended.c [hunk 36]
```diff
@@ -0,0 +1,1134 @@
...
+  // Record the undo for the actual move
+  if (marks_moved && undo == kExtmarkUndo) {
+    u_extmark_copy_place(buf, l_lnum, l_col, u_lnum, u_col, p_lnum, p_col);
+  }
+
+  return marks_moved;
+}
+
+// Get reference to line in kbtree_t, allocating it if neccessary.
+ExtMarkLine *extline_ref(buf_T *buf, linenr_T lnum, bool put)
```
Comment (justinmk) [inline src/nvim/mark_extended.c]: what is "extline"? These should be named `extmark_line_...`

**Review by justinmk — APPROVED** [review]

File: runtime/doc/api.txt [hunk 37]
```diff
@@ -438,6 +438,42 @@ Example: create a float with scratch buffer: >
     call nvim_win_set_option(win, 'winhl', 'Normal:MyHighlight')
 >
 
+==============================================================================
+Eextended marks						   *api-extended-marks*
```
Comment (ashfinal) [inline runtime/doc/api.txt]:
```suggestion
Extended marks						   *api-extended-marks*
```

File: runtime/doc/api.txt [hunk 38]
```diff
@@ -438,6 +438,42 @@ Example: create a float with scratch buffer: >
...
+=> [0, 2]
+
+Or we can look-up all marks in a buffer for our namespace (or by a range):
+`echo nvim_buf_get_extmarks(0, g:mark_ns, 0, -1, -1)`
+=> [[1, 0, 2]]
+
+Deleting the text all around an extended mark does not remove it. If you want
+to remove an extended mark, use the |nvim_buf_del_extmark()| function.
+
+The namepsace ensures your plugin doesn't have to deal with extmarks created
```
Comment (ashfinal) [inline runtime/doc/api.txt]:
```suggestion
The namespace ensures your plugin doesn't have to deal with extmarks created
```

File: runtime/doc/api.txt [hunk 39]
```diff
@@ -438,6 +438,42 @@ Example: create a float with scratch buffer: >
...
+by another plugin.
+
+Mark positions changed by an edit will be restored on undo/redo. Creating and
+deleting marks doesn't count as a buffer change on itself, i e new undo
```
Comment (ashfinal) [inline runtime/doc/api.txt]:
```suggestion
deleting marks doesn't count as a buffer change on itself, i.e. new undo
```

File: runtime/doc/api.txt [hunk 40]
```diff
@@ -1747,6 +1783,78 @@ nvim_buf_get_mark({buffer}, {name})                      *nvim_buf_get_mark()*
...
+                    (row, col) tuple or empty list () if extmark id was absent
+
+                                                     *nvim_buf_get_extmarks()*
+nvim_buf_get_extmarks({buffer}, {ns_id}, {start}, {end}, {amount})
+                List extmarks in a range (inclusive)
+
+                range ends can be specified as (row, col) tuples, as well as
+                extmark ids in the same namespace. In addition, 0 and -1 works
+                as shorthands for (0,0) and (-1,-1) respectively, so that all
+                marks in the buffer can be quieried as:
```
Comment (ashfinal) [inline runtime/doc/api.txt]:
```suggestion
                marks in the buffer can be queried as:
```

File: runtime/doc/api.txt [hunk 41]
```diff
@@ -1747,6 +1783,78 @@ nvim_buf_get_mark({buffer}, {name})                      *nvim_buf_get_mark()*
...
+nvim_buf_set_extmark({buffer}, {ns_id}, {id}, {line}, {col})
+                Create or update a namespaced mark at a position
+
+                If an invalid namespace is given, an error will be raised.
+
+                Parameters: ~
+                    {buffer}  The buffer handle
+                    {ns_id}   a identifier returned previously with
+                              nvim_create_namespace
+                    {id}      The extmark's id or 0 for next free id
```
Comment (ashfinal) [inline runtime/doc/api.txt]:
```suggestion
                    {id}      The extmark's id or 0 for next available id
```

File: runtime/doc/api.txt [hunk 42]
```diff
@@ -1747,6 +1783,78 @@ nvim_buf_get_mark({buffer}, {name})                      *nvim_buf_get_mark()*
...
+                    {row}     The row to set the extmark to.
+                    {col}     The column to set the extmark to.
+                    {opts}    Optional parameters. Currently not used.
+
+                Return: ~
+                    the nsmark_id for a new mark, or 0 for an update
```
Comment (ashfinal) [inline runtime/doc/api.txt]:
```suggestion
                    the namespaced extmark id for a new mark, or 0 for an update
```

**Review by ashfinal — CHANGES_REQUESTED** [review]
~~Some typos in docs.~~
Update: just wanted a peek of extmarks feature, saw some typos and felt the urge to correct them.
Feel free not to reply, reducing some noises :) @justinmk 
And thanks for your hint @bfredl

**Comment by justinmk** [comment]
> Some typos in docs.

Thanks, but I have a doc update pending anyway. There are many typos, but we want to avoid lots of typo comments in this PR.

**Comment by bfredl** [comment]
@ashfinal Thanks, but note that method-specific documentation in api.txt is generated, so some of those will be overwritten anyway. Method documentation should be reviewed in `src/nvim/api/*.c`.

File: src/nvim/mark_extended.h (hunk 27 above)
Comment (bfredl) [reply src/nvim/mark_extended.h]: I agree. This is one of the things I will consider after merge. IIRC there should only be two fundamental kinds of events: non-destructive adjustment after most edits (which just requires adjustment values), and destructive adjustment after delete (which requires saving a copy)

File: src/nvim/api/buffer.c [hunk 43]
```diff
@@ -999,6 +1003,222 @@ ArrayOf(Integer, 2) nvim_buf_get_mark(Buffer buffer, String name, Error *err)
...
+///
+/// @param buffer The buffer handle
+/// @param ns_id An id returned previously from nvim_create_namespace
+/// @param lower One of:  extmark id, (row, col) or 0, -1 for buffer ends
+/// @param upper One of: extmark id, (row, col) or 0, -1 for buffer ends
+/// @param amount Maximum number of marks to return or -1 for all marks found
+/// @param[out] err Details of an error that may have occurred
+/// @return [[nsmark_id, row, col], ...]
+Array nvim_buf_get_extmarks(Buffer buffer, Integer ns_id,
+                            Object start, Object end, Integer amount,
```
Comment (bfredl) [inline src/nvim/api/buffer.c]: `amount` param is probably very infrequent, we can replace it with `Dictionary opts` already (and in the future add flags for decorations etc).

**Comment by bfredl** [comment]
I added comments for the `Adjust*` structs, but did no changes now. Will try to reduce them later.

Also fixed the segfault on buffer wipe.

**Comment by bfredl** [comment]
Merged. Thanks @timeyyy for this great contribution.

As mentioned before, this just the start of "extmark" feature. Apart from doc/implementation improvements, an follow-up will be to integrate bufhl/virttext into it, which I hope to look into soon.

**Comment by timeyyy** [comment]
Thanks for getting this slotted in.

**Comment by justinmk** [comment]
> I have a doc update pending anyway.

https://github.com/neovim/neovim/pull/11376

**Pull request merged**